ccl/logictestccl: TestTenantLogic failed

[(ccl/logictestccl).TestTenantLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2444298&tab=buildLog) on [master@5d8f1d3dbc356573a1fb2d1d6bf564bf0370daff](https://github.com/cockroachdb/cockroach/commits/5d8f1d3dbc356573a1fb2d1d6bf564bf0370daff):

```
=== RUN   TestTenantLogic
    test_log_scope.go:155: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTenantLogic821109971
    test_log_scope.go:64: use -show-logs to present logs inline
--- FAIL: TestTenantLogic (445.64s)
=== RUN   TestTenantLogic/3node-tenant/schema_change_retry
=== PAUSE TestTenantLogic/3node-tenant/schema_change_retry
=== CONT  TestTenantLogic/3node-tenant/schema_change_retry
=== CONT  TestTenantLogic/3node-tenant/schema_change_retry
    logic.go:2400: 
        
        ../../sql/logictest/testdata/logic_test/schema_change_retry:54: 
        expected success, but found
        (40001) restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh): "sql txn" meta={id=73deac2e key=/Tenant/10/NamespaceTable/30/1/52/29/"t"/4/1 pri=0.02802829 epo=1 ts=1605412216.699754832,2 min=1605412215.154061480,0 seq=18} lock=true stat=PENDING rts=1605412215.980629186,1 wto=false max=1605412215.654061480,0
--- done: ../../sql/logictest/testdata/logic_test/schema_change_retry with config 3node-tenant: 11 tests, 1 failures
    logic.go:2908: 
        ../../sql/logictest/testdata/logic_test/schema_change_retry:54: error while processing
    logic.go:2908: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh): "sql txn" meta={id=73deac2e key=/Tenant/10/NamespaceTable/30/1/52/29/"t"/4/1 pri=0.02802829 epo=1 ts=1605412216.699754832,2 min=1605412215.154061480,0 seq=18} lock=true stat=PENDING rts=1605412215.980629186,1 wto=false max=1605412215.654061480,0
--- total progress: 9908 statements/queries
        --- FAIL: TestTenantLogic/3node-tenant/schema_change_retry (12.67s)
=== RUN   TestTenantLogic/3node-tenant
    --- FAIL: TestTenantLogic/3node-tenant (0.01s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestTenantLogic PKG=./pkg/ccl/logictestccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTenantLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/logictestccl).TestTenantLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2600828&tab=buildLog) on [master@cb97c6b6e4c4de09d00d021e223df0911f07127e](https://github.com/cockroachdb/cockroach/commits/cb97c6b6e4c4de09d00d021e223df0911f07127e):

```
            __auto__         {a}           1000       10              0
            __auto__         {a}           1000       10              0
            __auto__         {b}           1000       10              0
            __auto__         {b}           1000       10              0
            __auto__         {c}           1000       10              0
            __auto__         {c}           1000       10              0
            __auto__         {d}           1000       1               1000
            __auto__         {d}           1000       2               730
            
        but found (query options: "colnames,rowsort,retry" -> ignore the following ordering of rows) :
            statistics_name  column_names  row_count  distinct_count  null_count
            __auto__         {a,b,c}       1000       1000            0
            __auto__         {a,b,c}       1000       1000            0
            __auto__         {a,b,c}       1000       1000            0
            __auto__         {a,b}         1000       100             0
            __auto__         {a,b}         1000       100             0
            __auto__         {a,b}         1000       100             0
            __auto__         {a}           1000       10              0
            __auto__         {a}           1000       10              0
            __auto__         {a}           1000       10              0
            __auto__         {b}           1000       10              0
            __auto__         {b}           1000       10              0
            __auto__         {b}           1000       10              0
            __auto__         {c}           1000       10              0
            __auto__         {c}           1000       10              0
            __auto__         {c}           1000       10              0
            __auto__         {d}           1000       1               1000
            __auto__         {d}           1000       1               1000
            __auto__         {d}           1000       2               730
        goroutine 1202773 [running]:
        runtime/debug.Stack(0xc00c309278, 0x556e960, 0xc00e5e7660)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x574ad20, 0xc014141c80, 0xc00c309278)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x6d
        github.com/cockroachdb/cockroach/pkg/sql/logictest.(*logicTest).processSubtest(0xc0059fba00, 0x0, 0x0, 0xc00998c630, 0x0, 0xc003512bc0, 0x3f, 0x48e3fab, 0xc, 0x3, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:2249 +0x2267
        github.com/cockroachdb/cockroach/pkg/sql/logictest.(*logicTest).processTestFile(0xc0059fba00, 0xc003512bc0, 0x3f, 0x48e3fab, 0xc, 0x3, 0x0, 0x0, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:1794 +0x7b8
        github.com/cockroachdb/cockroach/pkg/sql/logictest.(*logicTest).runFile(0xc0059fba00, 0xc003512bc0, 0x3f, 0x48e3fab, 0xc, 0x3, 0x0, 0x0, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:3086 +0xd8
        github.com/cockroachdb/cockroach/pkg/sql/logictest.RunLogicTestWithDefaultConfig.func1.1(0xc014141c80)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:3285 +0x2de
        testing.tRunner(0xc014141c80, 0xc007785e80)
        	/usr/local/go/src/testing/testing.go:1123 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x2b3
--- done: ../../sql/logictest/testdata/logic_test/distsql_automatic_stats with config 3node-tenant: 10 tests, 0 failures
        --- FAIL: TestTenantLogic/3node-tenant/distsql_automatic_stats (47.84s)
=== RUN   TestTenantLogic/3node-tenant
    --- FAIL: TestTenantLogic/3node-tenant (1036.49s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestTenantLogic PKG=./pkg/ccl/logictestccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTenantLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/distsql_automatic_stats
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_automatic_stats
@@ -1,4 +1,4 @@
-# LogicTest: !metamorphic, !3node-tenant
+# LogicTest: !metamorphic
 
 # Disable automatic stats
 statement ok
@@ -18,9 +18,13 @@ INSERT INTO data SELECT a, b, c::FLOAT, NULL FROM
    generate_series(1, 10) AS b(b),
    generate_series(1, 10) AS c(c)
 
+# The query uses DISTINCT ON and ORDER BY to only show the latest statistic
+# available for each set of column names. This is important in order to
+# tolerate the rare case of multiple auto stats jobs running between two retry
+# iterations.
 query TTIII colnames,rowsort,retry
-SELECT statistics_name, column_names, row_count, distinct_count, null_count
-FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, distinct_count, null_count
+FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names, created
 ----
 statistics_name  column_names  row_count  distinct_count  null_count
 __auto__         {a,b,c}       1000       1000            0
@@ -40,8 +44,8 @@ UPDATE data SET d = 10 WHERE (a = 1 OR a = 2 OR a = 3) AND b > 1
 
 # There should be no change to stats.
 query TTIII colnames,rowsort
-SELECT statistics_name, column_names, row_count, distinct_count, null_count
-FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, distinct_count, null_count
+FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names ASC, created DESC
 ----
 statistics_name  column_names  row_count  distinct_count  null_count
 __auto__         {a,b,c}       1000       1000            0
@@ -60,21 +64,15 @@ statement ok
 UPDATE data SET d = 12 WHERE d = 10
 
 query TTIII colnames,rowsort,retry
-SELECT statistics_name, column_names, row_count, distinct_count, null_count
-FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, distinct_count, null_count
+FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names ASC, created DESC
 ----
 statistics_name  column_names  row_count  distinct_count  null_count
 __auto__         {a,b,c}       1000       1000            0
-__auto__         {a,b,c}       1000       1000            0
-__auto__         {a,b}         1000       100             0
 __auto__         {a,b}         1000       100             0
 __auto__         {a}           1000       10              0
-__auto__         {a}           1000       10              0
-__auto__         {b}           1000       10              0
 __auto__         {b}           1000       10              0
 __auto__         {c}           1000       10              0
-__auto__         {c}           1000       10              0
-__auto__         {d}           1000       1               1000
 __auto__         {d}           1000       2               730
 
 # Upsert more than 20% of the table.
@@ -85,27 +83,15 @@ generate_series(1, 10) AS b(b),
 generate_series(1, 5) AS c(c)
 
 query TTIII colnames,rowsort,retry
-SELECT statistics_name, column_names, row_count, distinct_count, null_count
-FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, distinct_count, null_count
+FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names ASC, created DESC
 ----
 statistics_name  column_names  row_count  distinct_count  null_count
-__auto__         {a,b,c}       1000       1000            0
-__auto__         {a,b,c}       1000       1000            0
 __auto__         {a,b,c}       1050       1050            0
-__auto__         {a,b}         1000       100             0
-__auto__         {a,b}         1000       100             0
 __auto__         {a,b}         1050       110             0
-__auto__         {a}           1000       10              0
-__auto__         {a}           1000       10              0
 __auto__         {a}           1050       11              0
-__auto__         {b}           1000       10              0
-__auto__         {b}           1000       10              0
 __auto__         {b}           1050       10              0
-__auto__         {c}           1000       10              0
-__auto__         {c}           1000       10              0
 __auto__         {c}           1050       10              0
-__auto__         {d}           1000       1               1000
-__auto__         {d}           1000       2               730
 __auto__         {d}           1050       3               365
 
 # Delete more than 20% of the table.
@@ -113,33 +99,15 @@ statement ok
 DELETE FROM data WHERE c > 5
 
 query TTIII colnames,rowsort,retry
-SELECT statistics_name, column_names, row_count, distinct_count, null_count
-FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, distinct_count, null_count
+FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names ASC, created DESC
 ----
 statistics_name  column_names  row_count  distinct_count  null_count
-__auto__         {a,b,c}       1000       1000            0
-__auto__         {a,b,c}       1000       1000            0
-__auto__         {a,b,c}       1050       1050            0
 __auto__         {a,b,c}       550        550             0
-__auto__         {a,b}         1000       100             0
-__auto__         {a,b}         1000       100             0
-__auto__         {a,b}         1050       110             0
 __auto__         {a,b}         550        110             0
-__auto__         {a}           1000       10              0
-__auto__         {a}           1000       10              0
-__auto__         {a}           1050       11              0
 __auto__         {a}           550        11              0
-__auto__         {b}           1000       10              0
-__auto__         {b}           1000       10              0
-__auto__         {b}           1050       10              0
 __auto__         {b}           550        10              0
-__auto__         {c}           1000       10              0
-__auto__         {c}           1000       10              0
-__auto__         {c}           1050       10              0
 __auto__         {c}           550        5               0
-__auto__         {d}           1000       1               1000
-__auto__         {d}           1000       2               730
-__auto__         {d}           1050       3               365
 __auto__         {d}           550        1               0
 
 # Test CREATE TABLE ... AS
@@ -149,8 +117,8 @@ CREATE TABLE copy AS SELECT * FROM data
 # Distinct count for rowid can be flaky, so don't show it. The estimate is
 # almost always 550, but occasionally it is 549...
 query TTII colnames,rowsort,retry
-SELECT statistics_name, column_names, row_count, null_count
-FROM [SHOW STATISTICS FOR TABLE copy] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, null_count
+FROM [SHOW STATISTICS FOR TABLE copy] ORDER BY column_names ASC, created DESC
 ----
 statistics_name  column_names  row_count  null_count
 __auto__         {a}           550        0
@@ -163,8 +131,8 @@ statement ok
 CREATE TABLE test_create (x INT PRIMARY KEY, y CHAR)
 
 query TTIII colnames,rowsort,retry
-SELECT statistics_name, column_names, row_count, distinct_count, null_count
-FROM [SHOW STATISTICS FOR TABLE test_create] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, distinct_count, null_count
+FROM [SHOW STATISTICS FOR TABLE test_create] ORDER BY column_names ASC, created DESC
 ----
 statistics_name  column_names  row_count  distinct_count  null_count
 __auto__         {x}           0          0               0
@@ -175,19 +143,14 @@ statement ok
 DELETE FROM copy WHERE true
 
 query TTII colnames,rowsort,retry
-SELECT statistics_name, column_names, row_count, null_count
-FROM [SHOW STATISTICS FOR TABLE copy] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, null_count
+FROM [SHOW STATISTICS FOR TABLE copy] ORDER BY column_names ASC, created DESC
 ----
 statistics_name  column_names  row_count  null_count
-__auto__         {a}           550        0
 __auto__         {a}           0          0
-__auto__         {b}           550        0
 __auto__         {b}           0          0
-__auto__         {c}           550        0
 __auto__         {c}           0          0
-__auto__         {d}           550        0
 __auto__         {d}           0          0
-__auto__         {rowid}       550        0
 __auto__         {rowid}       0          0
 
 
@@ -211,8 +174,8 @@ INSERT INTO my_schema.my_table SELECT k, NULL FROM
    generate_series(1, 10) AS k(k)
 
 query TTIII colnames,rowsort,retry
-SELECT statistics_name, column_names, row_count, distinct_count, null_count
-FROM [SHOW STATISTICS FOR TABLE my_schema.my_table] ORDER BY column_names::STRING, created
+SELECT DISTINCT ON (column_names) statistics_name, column_names, row_count, distinct_count, null_count
+FROM [SHOW STATISTICS FOR TABLE my_schema.my_table] ORDER BY column_names ASC, created DESC
 ----
 statistics_name  column_names  row_count  distinct_count  null_count
 __auto__         {k}           10         10              0

